gtsfm/retriever/netvlad_retriever.py

```diff
@@ -0,0 +1,175 @@
+            blocksize: size of matching sub-blocks when creating similarity matrix.
+        """
+        self._num_matched = num_matched
+        self._global_descriptor_model = GlobalDescriptorCacher(global_descriptor_obj=NetVLADGlobalDescriptor())
+        self._blocksize = blocksize
+        self._min_score = 0.5
+
+    def run(self, loader: LoaderBase, visualize: bool = True) -> List[Tuple[int, int]]:
+        """Compute potential image pairs.
+
+        Args:
+            loader: image loader. The length of this loader will provide the total number of images
+                for exhaustive global descriptor matching.
+
+        Return:
+            pair_indices: (i1,i2) image pairs.
+        """
+        num_images = len(loader)
+
+        sim = self.compute_similarity_matrix(loader, num_images)
+
+        query_names = loader.image_filenames()
+        # Avoid self-matching and disallow lower triangular portion
+        is_invalid_mat = ~np.triu(np.ones((num_images, num_images), dtype=bool))
+        np.fill_diagonal(a=is_invalid_mat, val=True)
+        pairs = pairs_from_score_matrix(
+            sim, invalid=is_invalid_mat, num_select=self._num_matched, min_score=self._min_score
+        )
+
+        if visualize:
+            plt.imshow(np.triu(sim.detach().cpu().numpy()))
+            os.makedirs(PLOT_SAVE_DIR, exist_ok=True)
+            plt.title("Image Similarity Matrix")
+            np.savetxt(
+                fname=os.path.join(PLOT_SAVE_DIR, "netvlad_similarity_matrix.txt"),
+                X=sim.detach().cpu().numpy(),
+                fmt="%.2f",
+                delimiter=",",
+            )
+            plt.savefig(os.path.join(PLOT_SAVE_DIR, "netvlad_similarity_matrix.jpg"), dpi=500)
+
+        named_pairs = [(query_names[i], query_names[j]) for i, j in pairs]
+        logger.info("Found %d pairs from the NetVLAD Retriever.", len(pairs))
+        logger.info("Image Name Pairs:" + str(named_pairs))
+        return pairs
+
+    def compute_similarity_matrix(self, loader: LoaderBase, num_images: int) -> torch.Tensor:
+        """Compute a similarity matrix between all pairs of images.
+
+        We use block matching, to avoid excessive memory usage.
+        We cannot fit more than 50x50 sized block into memory, on a 16 GB RAM machine.
+
+        A similar blocked exhaustive matching implementation can be found in COLMAP:
+        https://github.com/colmap/colmap/blob/dev/src/feature/matching.cc#L899
+
+        Returns:
+            sim: tensor of shape (num_images, num_images) representing similarity matrix.
+        """
+        device = "cuda" if torch.cuda.is_available() else "cpu"
+
+        if num_images > MAX_NUM_IMAGES:
+            raise RuntimeError("Cannot construct similarity matrix of this size.")
+
+        sim = torch.zeros((num_images, num_images))
+        num_blocks = math.ceil(num_images / self._blocksize)
+
+        for block_i in range(num_blocks):
+            for block_j in range(num_blocks):
```

Comment:
Can this be parallelized?

Reply:
It could be. One way would be to have a dask `.compute()` call right before we set up the SceneOptimizer's computational graph.

We could have the computation of each sub-block be a separate Delayed dask.

Reply:
@travisdriver I went ahead and converted this part to use Dask -- I can get a 10x + speedup on this section now with parallelization. Nice suggestion.